substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,830 @@
+						target,
+						stake_in_list,
+						calculated_stake,
+					);
+
+				return Err("target score in the target list is different than the expected".into())
+			}
+		}
+
+		frame_support::ensure!(
+			approvals_map.keys().count() == T::TargetList::iter().count(),
+			"calculated approvals count is different from total of target list.",
+		);
+
+		Ok(())
+	}
+
+	/// Try-state: checks if targets in the target list are sorted by score.
+	///
+	/// Invariant
+	///  * All targets in the target list are sorted by their score.
+	///
+	///  NOTE: unfortunatelly, it is not trivial to check if the sort correctness of the list if
+	///  the `SortedListProvider` is implemented by bags list due to score bucketing. Thus, we
+	///  leverage the [`SortedListProvider::in_position`] to verify if the target is in the
+	/// correct  position in the list (bag or otherwise), given its score.
+	#[cfg(feature = "try-runtime")]
+	pub fn do_try_state_target_sorting() -> Result<(), sp_runtime::TryRuntimeError> {
+		for t in T::TargetList::iter() {
+			frame_support::ensure!(
+				T::TargetList::in_position(&t).expect("target exists"),
+				"target list is not sorted"
+			);
+		}
+
+		Ok(())
+	}
+}
+
+impl<T: Config> OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	/// Fired when the stake amount of some staker updates.
+	///
+	/// When a nominator's stake is updated, all the nominated targets must be updated
+	/// accordingly.
+	///
+	/// Note: it is assumed that `who`'s staking ledger state is updated *before* this method is
+	/// called.
+	fn on_stake_update(
+		who: &T::AccountId,
+		prev_stake: Option<Stake<BalanceOf<T>>>,
+		stake: Stake<BalanceOf<T>>,
+	) {
+		// closure to calculate the stake imbalance of a staker.
+		let stake_imbalance_of = |prev_stake: Option<Stake<BalanceOf<T>>>,
+		                          voter_weight: ExtendedBalance| {
+			if let Some(prev_stake) = prev_stake {
+				let prev_voter_weight = Self::to_vote_extended(prev_stake.active);
+
+				if prev_voter_weight > voter_weight {
+					StakeImbalance::Negative(
+						prev_voter_weight.defensive_saturating_sub(voter_weight),
+					)
+				} else {
+					StakeImbalance::Positive(
+						voter_weight.defensive_saturating_sub(prev_voter_weight),
+					)
+				}
+			} else {
+				// if staker had no stake before update, then add all the voter weight
+				// to the target's score.
+				StakeImbalance::Positive(voter_weight)
+			}
+		};
+
+		if T::Staking::status(who)
+			.and(T::Staking::stake(who))
+			.defensive_proof(
+				"staker should exist when calling on_stake_update and have a valid status",
+			)
+			.is_ok()
+		{
+			let voter_weight = Self::weight_of(stake.active);
+
+			match T::Staking::status(who).expect("status checked above; qed.") {
+				StakerStatus::Nominator(nominations) => {
+					let _ = T::VoterList::on_update(who, voter_weight).defensive_proof(
+						"staker should exist in VoterList, as per the contract \
+                            with staking.",
```

Comment:
I have been trying to get your attention about this recently as well: https://github.com/paritytech/polkadot-sdk/issues/442#issuecomment-1927476773

This is not a breaking change of behavior. At least form the prespective of the THIS pallet, the statement of 

```
/// To keep the load off the chain as much as possible, changes made to the staked amount
/// via rewards and slashes are dropped and thus need to be manually updated through
/// extrinsics. In case of `bags-list`, this always means using `rebag` and `putInFrontOf`.
```

is entirely void. This pallet will ALWYAS update the score of a nominator when it is notified. 

If staking is still doing this trick where some updates to VoterList are dropped in the case of reward and slash, it is a matter to be documented elsewhere. In general, I wanted to already complain about the above doc comment that it is very off topic and it is explaining what "others pallet might do" instead of focusing on the point of view of this pallet. 

From the point of view of this pallet, both VoterList and TargetList are kept 100% up to date as far as I can see. The former based on the stake of a nominator, and the latter based on approval stakes.

Reply:
> is entirely void. This pallet will ALWYAS update the score of a nominator when it is notified.

Yes, that's correct. As per https://github.com/paritytech/polkadot-sdk/issues/442, I will keep the current logic, update the comments and add a try-state to check that the voter list is always sorted. Do you have any concerns with wrt keeping the voter list sorted at all times?

Reply:
It is only a matter of weight. How are we dealing with rewards right now? 

If a nominator is rewarded, we updated up to 16 notes in the `TargetList` for them (I couldn't find the code for this, but it is one of the important parts). IMO, one extra update to `VoterList` is no longer an issue if we handle 16 correctly.

Reply:
In other words `O(16) ~ O(17) ~ O(1)` 🙈

Reply:
Added a few thoughts on the overhead of keeping the target and voter lists sorted here https://hackmd.io/FH8Uhi2aQ5mD0aMm-BbqMQ. Keeping the voter list updated is negligible indeed.